src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -355,51 +354,64 @@ class EntityClient (requestContext: RequestContext, modelSchema: ModelSchema)(im
     }
   }
 
-  def importPFB(pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]],
-                workspaceNamespace: String, workspaceName: String, pfbRequest: PfbImportRequest): Future[PerRequestMessage] = {
+  def importPFB(workspaceNamespace: String, workspaceName: String, pfbRequest: PfbImportRequest, userInfo: UserInfo): Future[PerRequestMessage] = {
 
-    def callArrow = {
-      val gzipPipeline = addHeader (`Accept-Encoding`(gzip)) ~> pipeline ~> decode(Gzip)
-      gzipPipeline { Post(avroToRawlsURL, pfbRequest) }
-    }
+    // initial validation of the presigned url provided as an argument
+    def validatePfbUrl(pfbRequest: PfbImportRequest)(op: URL => Future[PerRequestMessage]) = {
+      val pfbUrl = try {
+        new URL(pfbRequest.url)
+      } catch {
+        case e: Exception => throw new FireCloudExceptionWithErrorReport(ErrorReport(BadRequest, s"Invalid URL: ${pfbRequest.url}", e))
+      }
+      val acceptableProtocols = Seq("https") //for when we inevitably change our mind and need to support others
 
-    def callRawls(entity: HttpEntity) = {
-      val authPipeline = authHeaders(requestContext) ~> pipeline
-      authPipeline {
-        Post(FireCloudDirectiveUtils.encodeUri(Rawls.entityPathFromWorkspace(workspaceNamespace, workspaceName) + "/batchUpsert"), entity)
+      if (!acceptableProtocols.contains(pfbUrl.getProtocol)) {
+        throw new FireCloudExceptionWithErrorReport(ErrorReport(BadRequest, "Invalid URL protocol: must be https only"))
       }
+
+      // TODO: add real (semantic) validation for origins, to ensure caller isn't supplying malice
+      op(pfbUrl)
     }
 
-    val pfbUrl = try {
-      new URL(pfbRequest.url)
-    } catch {
-      case e: Exception => throw new FireCloudExceptionWithErrorReport(ErrorReport(BadRequest, s"Invalid URL: ${pfbRequest.url}", e))
+    // probe Rawls with an empty upsert to this workspace. This checks that the workspace exists and that the user has
+    // permissions to perform an upsert.
+    def validateUpsertPermissions()(op: Boolean => PerRequestMessage) = {
+      val pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]] = authHeaders(requestContext) ~> sendReceive
+
+      val probeUpsert = pipeline {
+        Post(FireCloudDirectiveUtils.encodeUri(Rawls.entityPathFromWorkspace(workspaceNamespace, workspaceName) + "/batchUpsert"), List.empty[String])
```

Comment:
This is an interesting way of checking the user can upsert to a workspace. What's the reason behind doing it this way vs. just asking Sam if the user as permissions on the workspace?

Reply:
Sorry! I wrote this up in the PR description but I must have overwritten it in a second tab.

I decided to do it this way for a few reasons:
* encapsulation: it keeps Rawls as the only artifact that has knowledge of which permissions need to be checked to do an upsert. Otherwise, we'd have both Orch and Rawls implementing the same permission-check.
* future-proofing: if we ever change which permission needs to be checked, this call would still do the right thing. If we implemented the Sam check ourself, we'd have to update both Rawls and Orch.
* functionality: this call to Rawls validates that the workspace exists AND that the user has permission to upsert. If we queried Sam directly for permissions, we'd either have to make a second call to Rawls to verify the workspace exists, or skip that validation step.